nrm_django/grants/forms.py

```diff
@@ -0,0 +1,68 @@
+import datetime
+import uuid
+
+from django import forms
+from django.core.exceptions import ValidationError
+
+from .models import Grant
+
+
+class GrantForm(forms.ModelForm):
+    class Meta:
+        model = Grant
+        exclude = [
+            "application_id",
+        ]
+        widgets = {
+            "areas_effected": forms.Textarea(attrs={"cols": 80, "rows": 10}),
+        }
+
+    def clean_state_eo_date(self):
+        """
+        Ensures that if State EO Code is Yes, then a date must be entered.
+        """
+        cleaned_data = super().clean()
+        eo_code = cleaned_data.get("state_eo_code")
+        eo_date = cleaned_data.get("state_eo_date")
+
+        if eo_code == "Y" and not eo_date:
+            raise ValidationError(
+                "If this agreement is subject to state EO, you must enter an EO date."
+            )
+
+    def save(self, commit=True):
+        """
+        Ensures on save we generate and save several values:
+        * A new CN/PK
+        * Updated `status_date` if new or status has chanaged.
+        * `created_in_instance` and `modified_in_instance`
+
+        Note that the instance ID is currently hard-coded.
+        It may have no value beyond mimicking the legacy system,
+        but if it is needed we will still need to learn how to capture it.
+        """
+        instance = super().save(commit=False)
+        # first, we're gonna have to create our cn/pk
+        # * First character: Year
+        # * Second character: Region
+        # * Trailing characters: The instance created in
+        # * Middle characters: some unique ID
+        # TO-DO: figure out a better way to know the correct region. Maybe from contacts table?
+        year_char = datetime.date.today().strftime("%y")[1:]
+        short_uuid = str(uuid.uuid4())[:8]
+        instance_id = "10602"
```

Comment:
Could we comment or use a constant to convey that this is a magic value that happens to match lots of their current values?

Reply:
Yeah, I noted it a bit below but we can certainly make it more explicit. I also wonder if there's any real value in keeping the instance ID at all, but that's a conversation for another day.